src/gorilla.c

```diff
@@ -0,0 +1,261 @@
+        *runner |= data << idx64;
+        *(++runner) |= clearBits(data, dataLen) >> available;
+    }
+    *idx += dataLen;
+}
+
+// Read `dataLen` bits from `arr` at position `idx`
+static u_int64_t readBits(u_int64_t *arr, u_int64_t *idx, u_int8_t dataLen) {
+    u_int64_t rv = 0;
+    u_int8_t idx64 = *idx % _64BIT;
+    u_int8_t available = _64BIT - idx64;
+    u_int64_t *runner = &arr[*idx / _64BIT];
+
+    if(available >= dataLen) {
+        rv = clearBits(*runner >> idx64, dataLen);
+    } else {
+        u_int8_t left = dataLen - available;
+        rv = clearBits(*runner >> idx64, available);
+        rv |= clearBits(*++runner, left) << available;
+    }
+    *idx += dataLen;
+    return rv;
+}
+
+// returns `true` if `x` is <2^(nbits-1)-1 or >-2^(nbits-1)
+static bool bitRange(int64_t x, u_int8_t nbits) {
+    // x+1 since int2bin and bin2int are [-32, 31]
+	return -((1LL << (nbits - 1)) - 1) <= (x + 1) && 
+             (1LL << (nbits - 1))      >= (x + 1);
+}
+
+static bool isSpaceAvailable(CompressedChunk *chunk, u_int8_t size) {
+  u_int64_t available = (chunk->size * 8) - chunk->idx;
+  return (size <= available) ? true : false;
+}
+
+/***************************** APPEND ********************************/
+static int appendTS(CompressedChunk *chunk, u_int64_t timestamp) {
+  union64bits doubleDelta;
+
+  int64_t curDelta = timestamp - chunk->prevTimestamp;
+  assert(curDelta >= 0);
+  doubleDelta.i = curDelta - chunk->prevTimestampDelta;
+  // check if enough size exist
+  // if not return false
+
+  if (doubleDelta.i == 0) {
+    CHECKSPACE(chunk, 1 + 1); // CHECKSPACE adds 1 as minimum for double space
+    appendBits(chunk->data, &chunk->idx, 0x00, 1);
+  } else if (bitRange(doubleDelta.i, 7)) {
+    CHECKSPACE(chunk, 2 + 7 + 1);
+    appendBits(chunk->data, &chunk->idx, 0x01, 2);
+    appendBits(chunk->data, &chunk->idx, int2bin(doubleDelta.i, 7), 7);
+  } else if (bitRange(doubleDelta.i, 10)) {
+    CHECKSPACE(chunk, 3 + 10 + 1);
+    appendBits(chunk->data, &chunk->idx, 0x03, 3);
+    appendBits(chunk->data, &chunk->idx, int2bin(doubleDelta.i, 10), 10);
+  } else if (bitRange(doubleDelta.i, 13)) {
+    CHECKSPACE(chunk, 4 + 13 + 1);
+    appendBits(chunk->data, &chunk->idx, 0x07, 4);
+    appendBits(chunk->data, &chunk->idx, int2bin(doubleDelta.i, 13),13);
+  } else if (bitRange(doubleDelta.i, 16)) {
+    CHECKSPACE(chunk, 5 + 16 + 1);
+    appendBits(chunk->data, &chunk->idx, 0x0f, 5);
+    appendBits(chunk->data, &chunk->idx, int2bin(doubleDelta.i, 16),16);
+  } else {
+    CHECKSPACE(chunk, 5 + 64 + 1);
+    appendBits(chunk->data, &chunk->idx, 0x1f, 5);
+    appendBits(chunk->data, &chunk->idx, doubleDelta.u, 64);
+  }
+  chunk->prevTimestampDelta = curDelta;
+  chunk->prevTimestamp = timestamp;
+  return CC_OK;
+}
+
+static int appendV(CompressedChunk *chunk, double value) {
+  union64bits val;
+  val.d = value;
+  u_int64_t xorWithPrevious = val.u ^ chunk->prevValue.u;
+
+  // Checked for 1 bit in appendTS
```

Comment:
why check the +1 bit in appendTS instead of here?

Reply:
Since 1 is the minimum space required for the value.
If when writing the timestamp to the chunk there are no bits left, we will create a new one. If there is 1 bit, there is a chance (maybe even a good chance, depends on data), that it will be sufficient.
Otherwise, we will write both to the new chunk.